### DISCUSSION TITLE
How to grant AD Service Principal access to fetch metadata from Azure Function? "Contributor" is apparently insufficient!

### DISCUSSION THREAD

**Discussion by siegfried01** [community/community#67524] [Other Feature Feedback, Questions, & Ideas]
### Github workflow & RBAC for Azure AD service principals

Question

### Backgound:

This is a follow on question to [https://github.com/orgs/community/discussions/67346](url).

The first two REST calls seem to working inside my github workflow: [https://github.com/siegfried01/HelloAzureADAuthenticatedFunc/blob/master/.github/workflows/azure-github-workflow-REST-demo-powershell.yml#L44](url)

(I'm not sure about the second REST call to get the app function key because github blocks the output of the write-output command.. How can I verify that I'm getting the funciton key?)

### REST Call to fetch Metadata (Signatures) from Azure Function App Fails

However, I'm now trying fetch the meta data from my azure function. This works from my laptop:

    curl -X GET -H "Accept=application/json" -H "Authorization: Bearer $AZURE_HTTP_API_AUTH_TOKEN"  "https://management.azure.com/subscriptions/$SUBSCRIPTION_ID/resourceGroups/$RG/providers/Microsoft.Web/sites/$AZURE_FUNCTIONAPP_NAME/functions?api-version=2015-08-01"

This displays a list of functions in the function app when I execute it on my laptop.

The service principal in secrets.AZURE_CLIENT_SECRET was created with 

     $id=(az group show --name $rg --query 'id' --output tsv)
     az ad sp create-for-rbac --name $sp --sdk-auth --role contributor --scopes $id

Should this not be sufficient to fetch the meta data from my azure function? 

If not, I grant it reader access:

    New-AzRoleAssignment -ObjectId $spID -RoleDefinitionName "Reader" -Scope "/subscriptions/$subscriptionId/resourceGroups/$rg/providers/Microsoft.Web/sites/$AZURE_FUNCTIONAPP_NAME"

I went to my function app in the portal and confirmed that my service principal now has read access.

### Error Message

Granting Reader to the service principal seems to have had no affect: I'm still getting the same error:

    {"error":{"code":"AuthorizationFailed","message":"The client 'b3dcxxxx-xxxx-xxxx-xxxx-xxxxxxxx621d' with object id 'b3dcxxxx-xxxx-xxxx-xxxx-xxxxxxxx621d' does not have authorization to perform action 'Microsoft.Resources/subscriptions/resourceGroups/Microsoft.Web/functions/read' over scope '/subscriptions/acc2xxxx-xxxx-xxxx-xxxx-xxxx87bc27db/resourceGroups/providers/Microsoft.Web/sites' or the scope is invalid. If access was recently granted, please refresh your credentials."}}Script execution Complete

I don't understand why it says to refresh  my credentials. How do I do that? The stuff I pasted into github when creating secrets.AZURE_CLIENT_SECRET has not changed.

I tried doing 
    
    New-AzRoleAssignment -ObjectId $spID -RoleDefinitionName "Microsoft.Resources/subscriptions/resourceGroups/Microsoft.Web/functions/read" -Scope "/subscriptions/$subscriptionId/resourceGroups/$rg/providers/Microsoft.Web/sites/$AZURE_FUNCTIONAPP_NAME"

but that gave me the error Cannot find role definition with name 'Microsoft.Resources/subscriptions/resourceGroups/Microsoft.Web/functions/read'.

How can I make this work in a github workflow?

Thanks again!

Siegfried

**Comment by passadis** [comment]
Hello , i dont see an Action of List Function Keys in the Reader role . I think you have to either create a new Custom role with 
Other : List Web Apps Functions Keys 
List Function keys.

**Reply by siegfried01** [reply]
Please provide me some guidance on how to do this. A powershell example would be wonderful! 

Actually, fetching the function keys might be working fine. I cannot tell because github blocks my print statements. 

Regarding security: is it a problem if I succeed in displaying the function key inside the workflow logs in my public github repository? I'm think that I could just delete the azure function apps and recreate them. Would this safely create new function keys?

I am definitely failing when I attempt to fetch the operations (see above CURL command) with "AuthorizationFailed".

**Reply by siegfried01** [reply]
As per the other discussion [https://learn.microsoft.com/en-us/answers/questions/1355574/how-to-run-make-rest-calls-to-azure-resource-manag?page=1&orderby=Helpful&comment=answer-1318906#newest-answer-comment](url) I added "Website Contributor" and that includes getting the function keys and getting the operations.

Why am I getting the same old error?

**Reply by passadis** [reply]
Hello 
As per our discussion [https://learn.microsoft.com/en-us/answers/questions/1355574/how-to-run-make-rest-calls-to-azure-resource-manag](url) , in case the CORS setting does not help , i would like to recreate 
In case you agree on that either send me a clone of the repo with redacted values or send me the WorkFlow and update me on the variables to add on the repo
Regards

**Reply by siegfried01** [reply]
The project is public and you should be able to easily fork it.

However, I'm going to have create a new git branch and remove all the parameters for cosmosdb & Azure AD. I'll let you know when I have done that. In the mean time, see my other attempts here: https://learn.microsoft.com/en-us/answers/questions/1355574/how-to-run-make-rest-calls-to-azure-resource-manag


I'll let you know when the branch is ready.

Thanks

**Reply by passadis** [reply]
We are the two of us on both sides of this issue ,,,,,:)  msft and Git....
Ok let me know i could not find the repo or did not see it ...

**Comment by siegfried01** [comment]
Answer is here: [https://learn.microsoft.com/en-us/answers/questions/1355574/how-to-run-make-rest-calls-to-azure-resource-manag](url).
